encoding/ccf/encode_typedef.go

```diff
@@ -0,0 +1,252 @@
+/*
+ * Cadence - The resource-oriented smart contract programming language
+ *
+ * Copyright 2022-2023 Dapper Labs, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ccf
+
+import (
+	"fmt"
+
+	"github.com/onflow/cadence"
+)
+
+// encodeCompositeType encodes cadence.CompositeType in type definition as
+// language=CDDL
+// struct-type =
+//
+//	; cbor-tag-struct-type
+//	#6.160(composite-type)
+//
+// resource-type =
+//
+//	; cbor-tag-resource-type
+//	#6.161(composite-type)
+//
+// event-type =
+//
+//	; cbor-tag-event-type
+//	#6.162(composite-type)
+//
+// contract-type =
+//
+//	; cbor-tag-contract-type
+//	#6.163(composite-type)
+//
+// enum-type =
+//
+//	; cbor-tag-enum-type
+//	#6.164(composite-type)
+//
+// composite-type = [
+//
+//	id: id,
+//	cadence-type-id: cadence-type-id,
+//	fields: [
+//	    + [
+//	        field-name: tstr,
+//	        field-type: inline-type
+//	    ]
+//	]
+//
+// ]
+func (e *Encoder) encodeCompositeType(typ cadence.CompositeType, tids ccfTypeIDByCadenceType) error {
```

Comment:
I thought the same too and previously tried to include `tids ccfTypeIDByCadenceType` as part of encoder.  

However, it didn't fit in neatly with encoder API because:
- CCF IDs in `ccfTypeIDByCadenceType` is currently sequentially generated and is specific to encoded value,
- `Encoder` API allows encoding multiple values though `encoder.Encode()`.

So I made `tids ccfTypeIDByCadenceType` specific to encoded value, instead of encoder.

Reply:
Good point! 👌